File: Step 07 Recursion/Step 7.1 Get a Strong Hold/Sort a Stack/Sort a Stack.java


/*Complete the function below*/
class GfG{
    private Stack<Integer> insertEle(Stack<Integer> s, int x) {
        if (s.isEmpty() || x > s.peek()) {
            s.push(x);
            return s;
        }
        int temp = s.pop();
        insertEle(s, x);
        s.push(temp);
        return s;
    }
	public Stack<Integer> sort(Stack<Integer> s)
	{
		//add code here.
		if (!s.isEmpty()) {
		    int x = s.pop();
		    sort(s);
		    insertEle(s, x);
		}
		return s;
	}
}

File: Step 07 Recursion/Step 7.1 Get a Strong Hold/Sort a Stack/Sort a Stack.cpp


/*The structure of the class is
class SortedStack{
public:
	stack<int> s;
	void sort();
};
*/

/* The below method sorts the stack s 
you are required to complete the below method */
stack<int> insertEle(stack<int> s, int x) {
    if (s.empty() || x > s.top()) {
        s.push(x);
    } else {
        int temp = s.top();
        s.pop();
        s = insertEle(s, x);
        s.push(temp);
    }
    return s;
}
void SortedStack::sort() {
    if (!s.empty()) {
        int x = s.top();
        s.pop();
        sort();
        s = insertEle(s, x);
    }
}